R-package/R/callbacks.R

```diff
@@ -1,563 +1,923 @@
+#'
+#' Note that boosting might be interrupted before reaching this last iteration, for
+#' example by using the early stopping callback \link{xgb.cb.early.stop}.
+#'
+#' \item iteration Index of the iteration number that is being executed (first iteration
+#' will be the same as parameter `begin_iteration`, then next one will add +1, and so on).
+#'
+#' \item iter_feval Evaluation metrics for the `watchlist` that was supplied, either
+#' determined by the objective, or by parameter `feval`.
+#'
+#' For \link{xgb.train}, this will be a named vector with one entry per element in
+#' `watchlist`, where the names are determined as 'watchlist name' + '-' + 'metric name' - for
+#' example, if `watchlist` contains an entry named "tr" and the metric is "rmse",
+#' this will be a one-element vector with name "tr-rmse".
+#'
+#' For \link{xgb.cv}, this will be a 2d matrix with dimensions `[length(watchlist), nfolds]`,
+#' where the row names will follow the same naming logic as the one-dimensional vector
+#' that is passed in \link{xgb.train}.
+#'
+#' Note that, internally, the built-in callbacks such as \link{xgb.cb.print.evaluation} summarize
+#' this table by calculating the row-wise means and standard deviations.
+#'
+#' \item final_feval The evaluation results after the last boosting round is executed
+#' (same format as `iter_feval`, and will be the exact same input as passed under
+#' `iter_feval` to the last round that is executed during model fitting).
+#'
+#' \item prev_cb_res Result from a previous run of a callback sharing the same name
+#' (as given by parameter `cb_name`) when conducting training continuation, if there
+#' was any in the booster R attributes.
+#'
+#' Some times, one might want to append the new results to the previous one, and this will
+#' be done automatically by the built-in callbacks such as \link{xgb.cb.evaluation.log},
+#' which will append the new rows to the previous table.
+#'
+#' If no such previous callback result is available (which it never will when fitting
+#' a model from start instead of updating an existing model), this will be `NULL`.
+#'
+#' For \link{xgb.cv}, which doesn't support training continuation, this will always be `NULL`.
+#' }
+#'
+#' The following names (`cb_name` values) are reserved for internal callbacks:\itemize{
+#' \item print_evaluation
+#' \item evaluation_log
+#' \item reset_parameters
+#' \item early_stop
+#' \item save_model
+#' \item cv_predict
+#' \item gblinear_history
+#' }
+#'
+#' The following names are reserved for other non-callback attributes:\itemize{
+#' \item names
+#' \item class
+#' \item call
+#' \item params
+#' \item niter
+#' \item nfeatures
+#' \item folds
+#' }
+#'
+#' When using the built-in early stopping callback (\link{xgb.cb.early.stop}), said callback
+#' will always be executed before the others, as it sets some booster C-level attributes
+#' that other callbacks might also use. Otherwise, the order of execution will match with
```

Comment:
Maybe we need an internal callback that does this and split it up with the early stopping? Along with the comment for evaluation history instead of actual data, see `CallbackContainer` in Python.

Reply:
But if it's implemented as a callback, it'd still need to be executed before other callbacks.

Reply:
But isn't that how it's currently done? Only difference is that it's part of the public interface so the user can choose to pass it manually.